scipy/sparse/sparsetools/coo.h

```diff
@@ -258,4 +258,49 @@ void coo_matmat_dense(const npy_int64 nnz,
     }
 }
 
+
+template <class I, class T>
+void coo_matmat_dense_nd(const npy_int64 nnz,
+                         const npy_int64 n_dim,
+                         const npy_int64 n_col_B,
+                         const I shape_B[],
+                         const I shape_Y[],
+                         const I A_coords[], // flattened coordinates of shape (n_dim * nnz)
+                         const T Ax[],
+                         const T Bx[],
+                         T Yx[])
+{
+    std::vector<npy_int64> strides_B(n_dim);
+    std::vector<npy_int64> strides_Y(n_dim);
+
+    strides_B[n_dim - 1] = 1;
+    strides_Y[n_dim - 1] = 1;
+    for (npy_int64 i = n_dim - 2; i >= 0; --i) {
+        strides_B[i] = strides_B[i + 1] * shape_B[i + 1];
+        strides_Y[i] = strides_Y[i + 1] * shape_Y[i + 1];
+    }
```

Comment:
While we are looping through `dim` here, does it help to create a vector of `dim * nnz`? We recompute that `nnz` times in the inner loop and I don't think the compiler will optimize that. These kinds of things are often tricky so I'll leave it to you to figure out which is better.

I looked at maybe storing the `(n_dim-1) * nnz` `(ndim -2) * nnz` values but I think the compiler will get those because they don't depend on the loop index. And its not in the inner loop anyway.

Reply:
Would that mean computing something of this sort?
```cpp
    // Precompute offsets for A_coords[dim * nnz + n]
    std::vector<std::vector<I>> coord_offsets(n_dim, std::vector<I>(nnz));
    for (npy_int64 dim = 0; dim < n_dim; ++dim) {
        for (npy_int64 n = 0; n < nnz; ++n) {
            coord_offsets[dim][n] = A_coords[dim * nnz + n];
        }
```

Reply:
Sorry for not being more clear. That still has the double loop so `dim * nnz` gets executed n_dim*nnz times. I am trying to reduce the number of multiplications by moving the `dim * nnz` part outside of the loop that has an index not used in this operation. 

Of course we'll need to store values for all n_dim values that `dim` takes on, but that should be small. So, we need a third vector of length n_dim to store the `dim * nnz` values.  Suppose we call the vector `dim_nnz`.

Something like this maybe:
```python
    std::vector<npy_int64> strides_B(n_dim);
    std::vector<npy_int64> strides_Y(n_dim);
    std::vector<npy_int64> dim_nnz(n_dim);

    ... other stuff ...

    for (npy_int64 i = n_dim - 2; i >= 0; --i) {
        strides_B[i] = strides_B[i + 1] * shape_B[i + 1];
        strides_Y[i] = strides_Y[i + 1] * shape_Y[i + 1];
        dim_nnz[i] = i * nnz;
    }
```
This is `n_dim` new multiplications. But inside the inner loop of the double loop, we can refer to `dim_nnz[dim]` instead of computing and save `n_dim * nnz` multiplications (twice -- in two different lines).  Something like:
```python
                src_offset += A_coords[dim_nnz[dim] + n] * strides_B[dim];
                dst_offset += A_coords[dim_nnz[dim] + n] * strides_Y[dim];
```
And then, since it is already stored, you can use it twice more for the final adjustments with `A_coords[dim_nnz[n_dim - 2] + n]` in each offset update immediately after the inner loop. I guess one uses `dim_nnz[n_dim - 1]`.  Anyway, in total we replace n_dim*nnz + 2*nnz multiplications with n_dim multiplications and a bunch of lookups (which are fast).

Maybe the vector should be called something different. Your code comment in the example suggests `A_coords_offsets`. But I like `dim_nnz` (especially if you change from `i` to `dim` in the first loop.)

Reply:
Ah okay, got it now, that makes sense! i've addressed this now at 1e0e94aa175584783b9287b33cdb442ae223a57a.